File: spark-connector/chspark/src/main/scala/org/apache/spark/sql/ch/CHRaw.java
/*
 * Copyright 2017 PingCAP, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.ch;


public CHRaw {
	private static void dump(Field field, int order) {
		System.out.println("    #" + order + " name:\"" + field.getName() + "\" type:" +
			field.getType().getTypeID() + " nullable:" + field.isNullable());
	}

	private static void dump(Magic.Query query, boolean decode) throws Exception {
		Schema schema = query.schema();
		List<Field> fields = schema.getFields();
		int i = 0;
		if (decode) {
			System.out.println("[schema]");
			for (Field field: fields) {
				dump(field, i);
				i += 1;
			}
		}

		while (true) {
			VectorSchemaRoot block = query.next();
			if (block == null) {
				break;
			}
			if (!decode) {
				continue;
			}
			System.out.println("[result]");

			List<FieldVector> columns = block.getFieldVectors();
			int j = 0;
			for (FieldVector column: columns) {
				Field field = column.getField();
				dump(field, j);
				ValueVector.Accessor acc = column.getAccessor();
				for (int k = 0; k < acc.getValueCount(); ++k) {
					Object v = acc.getObject(k);
					if (v instanceof Character) {
						System.out.println("    " + (int)(Character)v);
					} else {
						System.out.println("    " + acc.getObject(k).toString());
					}
				}
				j += 1;
			}

			System.out.println("    ---");
		}

		System.out.println("[query done]");
	}

	public static void main(String[] args) throws Exception {
		if (args.length < 2) {
			System.out.println("usage: <bin> query ch-host [port]");
			System.exit(-1);
		}

		String query = args[0];
		String host = args[1];
		int port = 9001;
		if (args.length > 2) {
			port = Integer.parseInt(args[2]);
		}

		CHClient client = new CHClient(query, host, port);

		Schema schema = client.getSchema();
		dump(schema);

		for (client.hasNext()) {
			VectorSchemaRoot block = client.next();
			dump(block, true);
		}

		client.close();
	}
}

File: spark-connector/chspark/src/main/scala/org/apache/spark/sql/ch/CHRaw.scala
/*
 * Copyright 2017 PingCAP, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.ch


object CHRaw {
  def main(args: Array[String]) {
    if (args.size < 3) {
      println("usage: <bin> query-sql partitions conc [host] [port]")
      return;
    }

    val query: String = args(0)
    val partitions: Int = args(1).toInt
    val conc: Int = args(2).toInt

    val host: String = if (args.size >= 4) {
      args(3)
    } else {
      "127.0.0.1"
    }

    val port: Int = if (args.size >= 5) {
      args(4).toInt
    } else {
      9006
    }

    val workers = new Array[Thread](partitions);

    var totalRows: Long = 0

    def addRows(n: Long): Unit = {
      this.synchronized {
        totalRows += n
      }
    }

    for (i <- 0 until partitions) {
      workers(i) = new Thread {
        println("#" + i + " start")
        val resp = CHExecutorPool.get(query, host, port, "", conc, false)
        var block = resp.executor.next
        while (block != null) {
          val columns = block.decoded.block.getFieldVectors
          val rows: Long = if (columns.isEmpty) { 0 } else { columns.get(0).getAccessor().getValueCount }
          addRows(rows)
          block.close
          block = resp.executor.next
        }
        CHExecutorPool.close(resp)
        println("#" + i + " finish")
      }
    }

    workers.foreach(_.join)
    println("All finish, total rows: " + totalRows)
  }
}
